File: DSA/Java/src/dsa/lib/algo/sort/QuickSort.java
package dsa.lib.algo.sort;

import dsa.lib.Utils;

public class QuickSort {
    public void sort(int[] ary, int length) {
        quickSort(ary, 0, length - 1);
    }

    /**
     * QuickSort the array recursively, until the array has only one element.
     *
     * @param ary  The array.
     * @param low  The low endpoint of the array, inclusive.
     * @param high The high endpoint of the array, inclusive.
     */
    private void quickSort(int[] ary, int low, int high) {
        if (low >= high) return;

        int m = partition(ary, low, high);
        quickSort(ary, low, m - 1);
        quickSort(ary, m + 1, high);
    }

    /**
     * Partition the array into 2 sub-parts, return the pivot position.
     * All the elements on the left part are smaller than the pivot element,
     * all the elements on the right part are bigger than the pivot element.
     *
     * @param ary  The array.
     * @param low  The low endpoint of the array to be partitioned, inclusive.
     * @param high The high endpoint of the array to be partitioned, inclusive.
     * @return The pivot position.
     */
    private int partition(int[] ary, int low, int high) {
        int i = low;
        for (int j = low; j < high; ++j) {
            if (ary[j] < ary[high]) Utils.arrayswap(ary, i++, j);
        }
        Utils.arrayswap(ary, i, high);

        return i;
    }

}

File: DSA/Python/src/dsa/lib/algo/sort/quick_sort.py
from typing import List

from dsa.lib import utils


class QuickSort:
    def sort(self, ary: List[int]) -> None:
        n: int = len(ary)
        self.quick_sort(ary, 0, n - 1)

    def quick_sort(self, ary, low: int, high: int) -> None:
        """
        QuickSort the ary recursively, until the array has only one element.

        :param ary: The array.
        :param low: The low endpoint of the array, inclusive.
        :param high: The high endpoint of the array, inclusive.
        """
        if low >= high: return

        m: int = self.partition(ary, low, high)
        self.quick_sort(ary, low, m - 1)
        self.quick_sort(ary, m + 1, high)

    @staticmethod
    def partition(ary: List[int], low: int, high: int) -> int:
        """
        Partition the array into 2 sub-parts, return the pivot position.
        All the elements on the left part are smaller than the pivot,
        all the elements on the right part are bigger than the pivot.

        :param ary: The array.
        :param low: The low endpoint of the array to be partitioned, inclusive.
        :param high: The high endpoint of the array to be partitioned, inclusive.
        :return: The pivot position.
        """
        i: int = low
        for j in range(low, high):
            if ary[j] < ary[high]:
                utils.array_swap(ary, i, j)
                i += 1
        utils.array_swap(ary, i, high)

        return i
